tests/unit/test_charm.py

```diff
@@ -0,0 +1,310 @@
+        ]
+
+        remove_call_mocks = [self.patch(resource, "remove") for resource in all_resources]
+        self.patch_property(CephCsiCharm, "resources").return_value = all_resources
+        self.harness.set_leader(True)
+
+        self.harness.charm.purge_k8s_resources(MagicMock())
+
+        for remove_call in remove_call_mocks:
+            remove_call.assert_called_once()
+
+        self.assertEqual(self.harness.charm.unit.status.name, "blocked")
+        self.assertEqual(self.harness.charm.unit.status.message, "Missing relations: ceph")
+        self.assertFalse(self.harness.charm._stored.resources_created)
+
+    def test_resource_removal_missing_resources(self):
+        """Test that attempt to remove already removed resources does not cause exception."""
+        resource_name = "mock_resource"
+        all_resources = [StorageClass(MagicMock(), resource_name)]
+        self.patch(Resource, "remove").side_effect = ApiException(status=404)
+        logger_mock = self.patch(logger, "debug")
+        self.patch_property(CephCsiCharm, "resources").return_value = all_resources
+
+        self.harness.set_leader(True)
+
+        self.harness.charm.purge_k8s_resources(MagicMock())
+
+        logger_mock.assert_called_with("Resource %s is already removed.", resource_name)
+
+    def test_resource_removal_api_failure(self):
+        """Test that non-404 Api exceptions are re-raised when removing k8s resources."""
+        api_status = 401
+        api_reason = "Unauthenticated"
+        expected_exception = ApiException(status=api_status, reason=api_reason)
+        self.patch(Resource, "remove").side_effect = expected_exception
+        self.harness.set_leader(True)
+
+        with self.assertRaises(ApiException) as raised:
+            self.harness.charm.purge_k8s_resources(MagicMock())
+
+        self.assertEqual(raised.exception, expected_exception)
+
+    def test_ceph_join(self):
+        """Test procedure for ceph-mon joined relation."""
+        # This test is skipped because charm needs to implement logical flow
+        # add_relation -> add_unit_to_relation -> update_relation_data
+        # Currently it expects tht at least one unit has all the needed data at the time
+        # of RelationJoined event which is flow that's not supported by Operator testing
+        # framework.
+
+        # resources = [
+        #     {"kind": "Foo", "metadata": {"name": "resource1"}},
+        #     {"kind": "Bar", "metadata": {"name": "resource2"}},
+        # ]
+        # render_resources_mock = self.patch(CephCsiCharm, 'render_all_resource_definitions')
+        # render_resources_mock.return_value = resources
+        # create_resources_mock = self.patch(CephCsiCharm, 'create_ceph_resources')
+        #
+        # self.harness.set_leader(True)
+        # self.join_ceph_mon()
+        #
+        # render_resources_mock.assert_called_once()
+        # create_resources_mock.assert_called_once_with(resources)
```

Comment:
I think this test should be removed from the PR (and the `join_ceph_mon()` helper). Since there is another issue that tracks fixing the ceph-client relation implementation, and think that PR could include unit tests for ceph_join (and ceph_changed).